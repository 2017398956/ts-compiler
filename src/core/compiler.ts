import path from 'path';
import ts from 'typescript/lib/typescript';

export let contextModuleMap = {}

export type CompilerOptions = {
  compilerOptions?: ts.CompilerOptions
  context?: any
  transformers?: ts.CustomTransformers
}

/**
 * 编译ts代码字符串，输出js字符串
 */
export function compile(code: string, options?: CompilerOptions) {
  const context = options?.context
  const compilerOptions = options?.compilerOptions
  const transformers = options?.transformers

  contextModuleMap = typeof context === 'object' ? context : {}

  return new Promise<string>(async (resolve) => {
    // tsconfig 配置
    const options: ts.CompilerOptions = {
      module: ts.ModuleKind.CommonJS,
      target: ts.ScriptTarget.ES2015,
      jsx: ts.JsxEmit.React,
      ...(compilerOptions ?? {})
    };

    // 创建一个ts编译器读写器
    const compilerHost = ts.createCompilerHost(options);

    // 创建入口文件
    const uuid = setTimeout(() => { })
    const sourceFile = { fileName: `${uuid}.tsx`, sourceText: code, outputFileName: `${uuid}.js` }

    // 重写readFile
    const originalGetSourceFile = compilerHost.getSourceFile;
    compilerHost.getSourceFile = fileName => {
      if (fileName === sourceFile.fileName) {
        return ts.createSourceFile(fileName, code, options.target, true)
      } else {
        return originalGetSourceFile.call(compilerHost, fileName);
      }
    };

    // 将编译输出的文件写入moduleMap
    compilerHost.writeFile = (fileName, data) => {
      contextModuleMap[path.resolve(fileName).split('.')[0]] = data
      if (fileName === sourceFile.outputFileName) {
        resolve(data)
      }
    };

    // 创建编译器
    const program = ts.createProgram([sourceFile.fileName], options, compilerHost);

    // 输出结果
    program.emit(undefined, undefined, undefined, undefined, transformers);
  })
}

// 类型擦除不会进行类型检查
export const transpile = (code: string, options?: CompilerOptions) => ts.transpileModule(code, options)
