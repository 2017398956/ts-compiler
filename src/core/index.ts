export * from './compiler';
export * from './read-exports';
export * from './runCode';
export * from './transformer';
export * from './traverser';
