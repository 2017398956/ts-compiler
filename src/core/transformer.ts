import ts from 'typescript/lib/typescript';

/**
 * 提供一个transform helper函数
 */
export function visitNodes(
  node: ts.Node,
  context: ts.TransformationContext,
  callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node
): ts.SourceFile
export function visitNodes(
  node: ts.Node,
  context: ts.TransformationContext,
  callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node
): ts.Node {
  const newNode = callback(node, context);
  if (node !== newNode) {
    return newNode;
  }
  return ts.visitEachChild(node, childNode => visitNodes(childNode, context, callback), context);
}

export const createTransformer = (callback: (node: ts.Node, context: ts.TransformationContext) => ts.Node): ts.TransformerFactory<ts.SourceFile> =>
  context => node => visitNodes(node, context, callback)