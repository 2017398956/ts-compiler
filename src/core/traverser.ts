import ts from 'typescript/lib/typescript';

/**
 * 创建一个ast节点
 */
const createAstNode = (code: string) =>
  ts.createSourceFile(`${setTimeout(() => { })}`, code, ts.ScriptTarget.ES2015, /*setParentNodes */ true);

/**
 * 递归遍历ast所有节点
 */
const traverse = (node: ts.Node, callback: (node: ts.Node) => void) => {
  callback(node)
  ts.forEachChild(node, node => traverse(node, callback))
}

/**
 * 从一段代码创建一个ast并遍历节点
 */
export const traverseFromString = (code: string, callback: (node: ts.Node) => void) => {
  const root = createAstNode(code)
  traverse(root, callback)
}