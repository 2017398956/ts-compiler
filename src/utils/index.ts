export * from './get-import';
export * from './noexcept';
export * from './util';
export * from './walkFile';
