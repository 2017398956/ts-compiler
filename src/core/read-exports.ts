import fs from 'fs';
import ts from 'typescript/lib/typescript';
import { promisify } from 'util';
import vm from 'vm';

import { compile, context } from './compiler';

/**
 * 读取ts代码并执行，获取export的变量
 */
export async function readTsExport(code: string, compilerOptions?: ts.CompilerOptions) {
  const result = await compile(code, compilerOptions)
  // 创建执行上下文
  return vm.runInContext(result, vm.createContext(context))
}

/**
 * 读取ts文件并执行，获取export的变量
 */
export async function readTsFileExport(file: string, compilerOptions?: ts.CompilerOptions) {
  const buffer = await promisify(fs.readFile)(file)
  const code = buffer.toString()
  const result = await compile(code, compilerOptions)
  // 创建执行上下文
  return vm.runInContext(result, vm.createContext(context))
}