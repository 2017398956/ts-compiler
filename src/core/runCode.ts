import { dirname, resolve } from 'path';
import vm from 'vm';

import { contextModuleMap } from './compiler';

const cjsModule = {
  exports: {}, require: (id: string) => {
    const uri = resolve(id)
    if (uri in contextModuleMap) {
      return runCode(contextModuleMap[uri], uri)
    }
    return module.require(id)
  }
}

const createModule = (code: string) =>
  `;((module)=>{${code};exports.default=typeof exports.default=='undefined'?module.exports:exports.default;return module.exports;})({exports})`

/**
 * 执行一段js代码
 */
export const runCode = (code: string, __fileName?: string) => {
  return vm.runInNewContext(createModule(code), vm.createContext({
    ...cjsModule,
    ...(__fileName ? {
      __dirname: dirname(__fileName),
      __filename: __fileName
    } : {})
  }))
}