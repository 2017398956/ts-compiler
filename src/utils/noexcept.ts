import ts from 'typescript/lib/typescript';

import { createTransformer } from '../core';

/**
 * 函数添加noexcept注释，在编译阶段会自动添加try catch
 */
export const transformNoExcept = createTransformer(node => {
  // 判断是否是函数声明
  if (ts.isFunctionDeclaration(node)) {
    // 查找jsdoc标签
    const enable = !!ts.getJSDocTags(node).find(tag => {
      return tag.tagName.escapedText === 'noexcept'
    })
    // 如果使用了noexcept注释
    if (enable) {
      // 返回一个新的函数声明
      return ts.factory.createFunctionDeclaration(
        node.decorators,
        node.modifiers,
        node.asteriskToken,
        node.name,
        node.typeParameters,
        node.parameters,
        node.type, // 这以上函数签名都保持不变，下面只修改函数体部分
        ts.factory.createBlock([
          // 创建一个try catch语句
          ts.factory.createTryStatement(
            node.body, // try 中是函数体内容，下面创建catch语句部分
            ts.factory.createCatchClause('error', ts.factory.createBlock([
              ts.factory.createExpressionStatement(// 创建一个语句
                ts.factory.createCallExpression( // 创建调用语句，即console.log(error)
                  ts.factory.createIdentifier('console.log'), // 创建一个标识符（函数调用） 
                  [], // 类型参数，无
                  [ts.factory.createIdentifier('error')]) // 创建一个标识符（函数参数）
              )
            ])),
            undefined // finally就不创建了，只需要处理catch
          )
        ])
      )
    }
  }
  // 其他节点不处理，直接返回
  return node
})