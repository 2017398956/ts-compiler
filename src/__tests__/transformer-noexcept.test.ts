import ts from 'typescript/lib/typescript';

import { transpile } from '../core';
import { transformNoExcept } from '../utils/noexcept';

describe('Transform', () => {
  it('NoExcept', () => {
    const code = `
/**
 * @noexcept
 */
function main (): number {
  throw new Error()
}

main()
`
    const result = transpile(code, {
      transformers: {
        before: [
          transformNoExcept
        ]
      }
    })

    expect(result.outputText).toMatchSnapshot()
  })
})