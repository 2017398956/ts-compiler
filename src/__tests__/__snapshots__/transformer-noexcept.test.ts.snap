// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform NoExcept 1`] = `
"function main() { try {
    throw new Error();
}
catch (error) {
    console.log(error);
} }
main();
"
`;
